### ISSUE TITLE
Improve Bless/Curse Manager

### ISSUE THREAD

**Issue by Entrox-Licher** [argonui/SCED#516]
The Bless/Curse Manager has several areas that I feel could be improved upon, and given the upcoming influx of bless/curse cards, it will be good to tackle these before Feast releases (on the mod that is).

- Tracking of tokens is handled manually through the buttons.  This can lead to situations where the manager thinks there are different amounts of tokens in play than there actually are.  I'd recommend tracking this automatically by tracking when tokens enter/exit the chaos bag or the individual token storage bags (the ones next to the chaos bag, which also aren't being used by the manager)
- Sealing could be improved.  Currently for cards that don't have context menu support, the Take/Return option exists, but this works entirely sequentially, which will often result in returning the wrong token when returning (especially when sealing tokens to enemies or other scenario cards).  I suggest adding context menu support to the tokens themselves, allowing the ability to right click and return them to the bag or supply, and properly update the Manager accordingly.  This could also replace large chunks of the `CardsThatSealsTokens` logic too.  Alternatively, having events in place so the player can manually deal with the tokens (returning to player bag or supply or deleted or put into trash).  Most of this would be handled automatically if the first bullet were implemented.
- Tracking Bless/Curse tokens in play is a bit cumbersome.  It is good we have a GameKey for it, but with how little information is needed to be displayed, I don't think it would be hard to add that as a small section to the Manager itself, so it is always visible at a glance.
- Finally, the Manager is located in a really awkward position, away from all other objects that interact with the chaos bag.  I would recommend moving it either up near the chaos bag itself or near the Token Arranger & Chaos Bag Manager

These are all the main issues that come to mind at the moment, I'll update this issue thread with more if they occur to me.  Cheers!

**Comment by Chr1Z93** [comment]
1) I don't really understand the first point, except that there is not tracking of manually edits to the chaos bag. My takeaway would be to remove the additional bless / curse containers since they are obsolete due to the manager and it's so long in use that it's highly likely to not break.

2) I agree that a context menu to return a token directly on the tokens would be useful. The "Take/Return" option is mostly for legacy support - I don't see any reason to improve this since no one should be using this at all.

3) Very good point! I think we could just have a small XML UI that is either always visible or visible onHover of the manager.

4) I agree that the position is kind of awkward, but the main advantage is that this is central. I don't like moving it outside of the main playarea.

**Comment by Glitchy-Tozier** [comment]
> Finally, the Manager is located in a really awkward position, away from all other objects that interact with the chaos bag. I would recommend moving it either up near the chaos bag itself or near the Token Arranger & Chaos Bag Manager

While I agree that a position near the chaos bag would quite intuitive, I think the most important thing is to keep the bless/curse manager easily accessible.

I think the main drawback atm is that the bless/curse manager isn't that accessible to players red and green.

An option could be to have some small UI near the "draw chaos token" button. I'm not 100% convinced of that, though.
`+` Intuitive position
`+` Handy for removing bless/curse tokens after drawing them
`-` Takes up important space
`-` Removes the important aspect of "I see someone else's hand hovering over the global b/c manager, meaning I won't have to add/remove the tokens".

**Comment by Chr1Z93** [comment]
> > Finally, the Manager is located in a really awkward position, away from all other objects that interact with the chaos bag. I would recommend moving it either up near the chaos bag itself or near the Token Arranger & Chaos Bag Manager
> 
> While I agree that a position near the chaos bag would quite intuitive, I think the most important thing is to keep the bless/curse manager easily accessible.
> 
> I think the main drawback atm is that the bless/curse manager isn't that accessible to players red and green.
> 
> An option could be to have some small UI near the "draw chaos token" button. I'm not 100% convinced of that, though. `+` Intuitive position `+` Handy for removing bless/curse tokens after drawing them `-` Takes up important space `-` Removes the important aspect of "I see someone else's hand hovering over the global b/c manager, meaning I won't have to add/remove the tokens".

Thanks for the idea! What do you think of a global UI (at the edge of your screen) for handling bless / curses? Could be collapsible and only show in full size when hovered

**Comment by Glitchy-Tozier** [comment]
> Thanks for the idea! What do you think of a global UI (at the edge of your screen) for handling bless / curses? Could be collapsible and only show in full size when hovered

I thought about it, and there's some interesting pro & con arguments. Unfortunately, I think the cons outweigh the pros:

#### Pro

It's comfortable. Wherever you are on the table, you have instant control over bless & curse tokens.
Additionally, it seems like we could expand on that Idea and and move all of the chaos-token stuff there. It sounds absolutely sick being able to hover a button on the side of your screen and seeing what tokens currently are in the bag, and being able to add/remove tokens. (basically moving the _token arranger_ and _token manager_ in there. They both seem kind of out of place anyway at the moment)

#### Con

I feel like this change would work against another core design philosophy of SCED.
To me it seems like the current structure is to have everything on the side of the screen matter _only_ when setting up the campaign (and sometimes in-between scenarios)
- Downloading the campaign (you'll be using for the rest of the campaign)
- Downloading special investigators (you'll be using for the rest of the campaign)
- Setting up helpers (you'll be using for the rest of the campaign)
- Sometimes you might add helpers in-between scenarios for example if someone decides to get a Backpack(2)

This is why I feel the current design makes you use the buttons on the side of your screen only for setting up the campaign, and then wants you to forget about them while actually playing.
Moving everything token-related over there means making those buttons relevant mid-scenario, which is different than the established design.

---

Tbh, I see both sides being strong arguments. What do you think?

**Comment by Chr1Z93** [comment]
I agree with your points if this was about moving all token related functionality into a separate UI. But as you explained, this falls flat since it would be a) a bunch of additional information and b) not something you need all the time. So I'd like to ask you to rethink your stance about this matter for just the bless/curse manager (since you actually **need** it during gameplay. The UI could feature a "burger-menu" like button (see doom counter's "Reset" button) or simply "minimize" if it is not hovered, so that it is not obtrusive when you don't need it.

**Comment by Glitchy-Tozier** [comment]
> I agree with your points if this was about moving all token related functionality into a separate UI. But as you explained, this falls flat since it would be a) a bunch of additional information and b) not something you need all the time. So I'd like to ask you to rethink your stance about this matter for just the bless/curse manager (since you actually **need** it during gameplay. The UI could feature a "burger-menu" like button (see doom counter's "Reset" button) or simply "minimize" if it is not hovered, so that it is not obtrusive when you don't need it.

I don't quite understand what you're trying to say, could you rephrase it?
Are you saying that despite those arguments, you'd like to move the bless/curse manager into the floating buttons?

**Comment by Chr1Z93** [comment]
In my opinion your arguments were for a different situation and don't apply to this. I don't think transforming the Bless/Curse manager into an UI instead of a stationary object would break any current design principles.

**Comment by Glitchy-Tozier** [comment]
As far as I understood your proposal was to make it one of the floating buttons, the same positioning as the download-community-content-button and the button that lets you load helpers, correct?
If so, the con-argument still applies. We would be adding something in that spot that regularly gets used mid-scenario. Which is contrary to the current usage of those elements.

**Comment by Chr1Z93** [comment]
Ah, we're having a misunderstanding. This would not be a floating button, but rather a small "window" (more like the navigation overlay).

**Comment by Glitchy-Tozier** [comment]
My argument was targeted at anything floating, which iirc the navigation overlay did. It was that currently floating things that are "stuck to the screen" (as opposed to the table) are there to be used before campaigns/scenarios, but to be forgotten mid-scenario.

Also, irrespective of what we're planning, I realized I enjoy the current bless/curse manager :)
It's nice and central, and it allows you to do anything you want with using a minimal amount of clicks. That's really important, especially with even more bless/curse cards coming out.

Another random idea:
Numpad 1&2 currently aren't getting used, right?
What is your opinion on
```
Numpad1 → Add Bless
Shift + Numpad1 → Remove Bless
Numpad2 → Add Curse
Shift + Numpad2 → Remove Curse
```

**Comment by Chr1Z93** [comment]
Sadly, we can't use any modifiers for hotkeys (since TTS doesn't support that).

**Comment by Chr1Z93** [comment]
Need to add:
- proper handling for manually (or by discard function) returned tokens
- "Return 3" option for Nephthys (and possibly other cards?)
- rename "Wendy's Menu" to something more specific
